cli/register/servicewait.go

```diff
@@ -24,115 +105,193 @@ func WaitForService(org string, waitService string, waitTimeout int, pattern str
+				if len(s.ServiceVersions) > 0 && CheckService(s.ServiceURL, s.ServiceOrg, nodeArch, s.ServiceVersions[0].Version, nodeType, userOrg, userPw) {
+					servSpecArr = append(servSpecArr, serviceSpec{name: s.ServiceURL, org: s.ServiceOrg, status: 0, serviceUp: 0})
+				}
+			}
+		}
+	} else {
+		servSpecArr = append(servSpecArr, serviceSpec{name: waitService, org: org, status: 0, serviceUp: 0})
+	}
+
 	// 1. Wait for the /service API to return a service with url that matches the input
 	// 2. While waiting, report when at least 1 agreement is formed
 
-	msgPrinter.Printf("Waiting for up to %v seconds for service %v/%v to start...", waitTimeout, org, waitService)
+	msgPrinter.Printf("Waiting for up to %v seconds for following services to start...", waitTimeout)
 	msgPrinter.Println()
 
+	for _, ss := range servSpecArr {
+		msgPrinter.Printf("\t%v/%v", ss.org, ss.name)
+		msgPrinter.Println()
+
+	}
 	// Save the most recent set of services here.
 	services := api.AllServices{}
 
 	// Start monitoring the agent's /service API, looking for the presence of the input waitService.
 	updateCounter := UpdateThreshold
-	serviceUp := 0
-	serviceFailed := false
 	now := uint64(time.Now().Unix())
-	for (uint64(time.Now().Unix())-now < uint64(waitTimeout) || serviceUp > 0) && !serviceFailed {
+	for uint64(time.Now().Unix())-now < uint64(waitTimeout) {
 		time.Sleep(time.Duration(3) * time.Second)
 		if _, err := cliutils.HorizonGet("service", []int{200}, &services, true); err != nil {
 			cliutils.Fatal(cliutils.CLI_GENERAL_ERROR, err.Error())
 		}
 
 		// Active services are services that have at least been started. When the execution time becomes non-zero
-		// it means the service container is started. The container could still fail quickly after it is started.
+		// it means the servie container is started. The container could still fail quickly after it is started.
 		instances := services.Instances["active"]
-		for _, serviceInstance := range instances {
-
-			if !(serviceInstance.SpecRef == waitService && serviceInstance.Org == org) {
-				// Skip elements for other services
-				continue
-
-			} else if serviceInstance.ExecutionStartTime != 0 {
-				// The target service is started. If stays up then declare victory and return.
-				if serviceUp >= ServiceUpThreshold {
-					msgPrinter.Printf("Service %v/%v is started.", org, waitService)
-					msgPrinter.Println()
-					return
-				}
 
-				// The service could fail quickly if we happened to catch it just as it was starting, so make sure
-				// the service stays up.
-				serviceUp += 1
+		// for each service that we are monitoring, check if they are up
+		for i, ss := range servSpecArr {
+			if ss.status == 0 {
+				for _, serviceInstance := range instances {
+					if serviceInstance.SpecRef == ss.name && (serviceInstance.Org == ss.org || ss.org == "*") {
+						if ss.org == "*" {
+							if ServiceInserted(servSpecArr, ss.name, serviceInstance.Org) {
```

Comment:
Can you explain what's going on in this code, I'm not following it. You're modifying the current array elements and then appending the current element to the end....

Reply:
This is the policy case where org is '*'. Since we do not know how many agreements there will be, we insert the ones that match into the beginning of the list and add the element with '*' as org to the end so we can keep adding new agreements that match. After the loops are done, we can simply remove the last element in the array (the '*' org).

Reply:
But in the policy case, -s "\*" is not supported, the user has to provide a service name. Org is optional in this case. So there is only 1 service that we need to monitor for.

Reply:
Fixed it. I removed the need for the serviceInstance function and the place where we removed the last element of servSpecArr. In this case, we simply need to catch one that matches.